src/Channels.Networking.Sockets/SocketConnection.cs

```diff
@@ -0,0 +1,304 @@
+﻿using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Channels.Networking.Sockets
+{
+    public class SocketConnection : IChannel
+    {
+        private Socket _socket;
+        private Channel _input, _output;
+        private ChannelFactory _channelFactory;
+        private readonly bool _ownsChannelFactory;
+
+        static readonly EventHandler<SocketAsyncEventArgs> _asyncCompleted = OnAsyncCompleted;
+
+        public IReadableChannel Input => _input ?? CreateInput();
```

Comment:
You can use this https://github.com/davidfowl/Channels/blob/master/src/Channels/Channel.cs#L46. The concrete channel impl exposes a task you can use to wait on before you start writing data to the input. Example here https://github.com/davidfowl/Channels/blob/master/src/Channels/ChannelFactory.cs#L38